models/resnet/resnet.hpp

```diff
@@ -0,0 +1,554 @@
+   * @param outSize Number of output maps.
+   * @param numBlocks Number of layers to create.
+   * @param stride Single parameter for StrideHeight and strideWidth.
+   */
+  void MakeLayer(const std::string& block,
+                 const size_t outSize,
+                 const size_t numBlocks,
+                 const size_t stride = 1)
+  {
+    bool downSample = false;
+
+    if (block == "basicblock")
+    {
+      if (stride != 1 || downSampleInSize != outSize * basicBlockExpansion)
+        downSample = true;
+      BasicBlock(downSampleInSize, outSize * basicBlockExpansion, stride,
+          stride, downSample);
+      downSampleInSize = outSize * basicBlockExpansion;
+      for (size_t i = 1; i != numBlocks; ++i)
+        BasicBlock(downSampleInSize, outSize);
+    }
+
+    else if (block == "bottleneck")
+    {
+      if (stride != 1 || downSampleInSize != outSize * bottleNeckExpansion)
+        downSample = true;
+      BottleNeck(downSampleInSize, outSize, stride, stride, downSample);
+      downSampleInSize = outSize * bottleNeckExpansion;
+      for (size_t i = 1; i != numBlocks; ++i)
+        BottleNeck(downSampleInSize, outSize);
+    }
+  }
+
+  /**
+   * Return the convolution output size.
+   *
+   * @param size The size of the input (row or column).
+   * @param k The size of the filter (width or height).
+   * @param s The stride size (x or y direction).
+   * @param padding The size of the padding (width or height) on one side.
+   * @return The convolution output size.
+   */
+  size_t ConvOutSize(const size_t size,
+                     const size_t k,
+                     const size_t s,
+                     const size_t padding)
+  {
+    return std::floor((size - k + 2 * padding) / s) + 1;
+  }
+
+  //! Locally stored DarkNet Model.
+  ann::FFN<OutputLayerType, InitializationRuleType> resNet;
+
+  //! Locally stored number of channels in the image.
+  size_t inputChannel;
+
+  //! Locally stored width of the image.
+  size_t inputWidth;
+
+  //! Locally stored height of the image.
+  size_t inputHeight;
+
+  //! Locally stored number of output classes.
+  size_t numClasses;
+
+  //! Locally stored width of image for downSample block.
+  size_t downSampleInputWidth;
+
+  //! Locally stored height of image for downSample block.
+  size_t downSampleInputHeight;
+
+  //! Locally stored expansion for BasicBlock.
+  size_t basicBlockExpansion = 1;
+
+  //! Locally stored expansion for BottleNeck.
+  size_t bottleNeckExpansion = 4;
+
+  //! InSize for ResNet block creation.
+  size_t downSampleInSize = 64;
+
+  //! Locally stored map to constructor different ResNet versions.
+  std::map<size_t, std::map<std::string, std::array<size_t, 4>>> ResNetConfig =
+      {
+        {18, {{"basicblock", {2, 2, 2, 2}}}},
+        {34, {{"basicblock", {3, 4, 6, 3}}}},
+        {50, {{"bottleneck", {3, 4, 6, 3}}}},
+        {101, {{"bottleneck", {3, 4, 23, 3}}}},
+        {152, {{"bottleneck", {3, 8, 36, 3}}}}
+      };
```

Comment:
Should we add a function to add custom blocks (getter / setter) so that user can have a custom resnet if needed.

Reply:
I don't know if that makes sense because pre trained models or specific models have a specific architecture so it doesn't makes sense to me add such a functionality here, users can use mlpack to create custom models.